include/adm/detail/id_parser.hpp

```diff
@@ -23,19 +23,32 @@ namespace adm {
       void check_size(size_t size) {
         if (id.size() != size) {
           std::ostringstream errorString;
-          errorString << "invalid " << type << " (wrong length): " << id;
+          errorString << "invalid " << type << " (wrong length, should be "
+                      << std::to_string(size) << " characters): " << id;
           throw std::runtime_error(errorString.str());
         }
       }
 
+      void check_prefix(const char *prefix, size_t size) const {
+        assert(prefix && prefix[size] == 0);
```

Comment:
this results in an unused parameter warning in non-debug builds

I'm not sure about this change -- the original intent was to avoid allocating a std::string for every ID that is parsed, and this undoes that, while also having the length passed for no reason.

Perhaps it would be better to just use `check_prefix(std::string const &prefix)`, and put the prefixes somewhere as static `std::string`s?

Reply:
I got a bit nerd-sniped by this. b0ab49afe51d98e14733f85992f08bb22858ddf9 and 1a3f6dd4dc3c77cff087beaddefe8dd06a53f57d rework the id parsing. The first one was just an attempt at 'putting the prefixes somewhere', which led on to the second one via 'hey we could just..'

Let's you declare the id formatting via a couple of traits classes then let the parser work out what to do with that.
I think the end result makes the ID classes look nicer, with less duplication and a consistent order for validation/parsing, at the expense of making the id_parser more complicated. Not sure if it's a good thing or not.

Reply:
https://github.com/ebu/libadm/blob/1a3f6dd4dc3c77cff087beaddefe8dd06a53f57d/src/elements/audio_block_format_id.cpp#L105-L138 
is an example of what I mean. 

The ID benchmarks show the same result as on master for parsing (within the variation I get between runs), and slightly better for formatting. Doesn't significantly impact performance either way.

Reply:
Seems reasonable, though it's quite a bit more code.

My ultra-simple idea was to base everything on the template/format string at run-time -- like this but without templates -- you can check the length, prefix/underscores (by checking that all non-lowercase letters match), and use the placeholder letters to parse/fill in hex values.

```cpp
// somewhere global
const std::sting format =  "AB_yyyyxxxx_zzzzzzzz";

// parse
check_format(format, id); // checks length, AB_ and second _
int index = parse_hex(format, id, 'z'); // extracts z bit, repeat for y, x, z

// format
std::sting format =  "AB_yyyyxxxx_zzzzzzzz";
std::strng id = format;
fill_hex(format, id, 'x', value) // fills in x part
```

Technically this means iterating over the strings a few more times, but maybe that's better than a lot of templates? The performance isn't really a problem, it just needs to not be very slow (like the regex-and-iostreams version).

This version is neat and looks good to me though!